bundle/src/main/java/com/adobe/acs/commons/rewriter/impl/TemplatedTransformer.java

```diff
@@ -0,0 +1,116 @@
+import com.adobe.acs.commons.properties.util.TemplateReplacementUtil;
+import com.adobe.acs.commons.rewriter.ContentHandlerBasedTransformer;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.rewriter.ProcessingComponentConfiguration;
+import org.apache.sling.rewriter.ProcessingContext;
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.AttributesImpl;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * {@link org.apache.sling.rewriter.Transformer} used to process HTML requests and replace content tokens found in the
+ * rendered HTML.
+ */
+public class TemplatedTransformer extends ContentHandlerBasedTransformer {
+
+    private Map<String, Object> properties;
+    private PropertyAggregatorService aggregatorService;
+
+    public TemplatedTransformer() {
+    }
+
+    public TemplatedTransformer(PropertyAggregatorService propertyAggregatorService) {
+        this.aggregatorService = propertyAggregatorService;
+    }
+
+    @Override
+    public void init(ProcessingContext processingContext, ProcessingComponentConfiguration processingComponentConfiguration) throws IOException {
+        SlingHttpServletRequest request = processingContext.getRequest();
+
+        properties = aggregatorService.getProperties(request);
+    }
+
+    public void startElement(String uri, String localName, String quaName, Attributes atts) throws SAXException {
+        if (shouldRun() && localName.equals("a")) {
+            AttributesImpl newAttrs = new AttributesImpl(atts);
+            for (int i = 0; i < newAttrs.getLength(); i++) {
+                String currentAttribute = newAttrs.getValue(i);
+                if (TemplateReplacementUtil.hasPlaceholder(currentAttribute)) {
+
+                    // Get the current placeholder in the string
+                    String placeholder = TemplateReplacementUtil.getPlaceholder(currentAttribute);
+
+                    // Transform it to the key in the property map
+                    String key = TemplateReplacementUtil.getKey(placeholder);
+
+                    // If the placeholder key is in the map then replace it
+                    if (properties.containsKey(key)) {
+                        String replaceValue = (String) properties.get(key);
+                        newAttrs.setValue(i, currentAttribute.replace(placeholder, replaceValue));
+                    }
+                }
+            }
+            getContentHandler().startElement(uri, localName, quaName, newAttrs);
+        } else {
+            getContentHandler().startElement(uri, localName, quaName, atts);
+        }
+    }
+
+    public void characters(char[] ch, int start, int length) throws SAXException {
+        String currentString = new String(ch);
+        int chunkLength = length;
+
+        if (shouldRun()) {
+            String currentChunk = currentString.substring(start, start + length);
```

Comment:
Is it possible to be lucky/unlucky and have a placeholder that spans the end of one chunk and the start of the next?

Reply:
I've really tried to have this happen but I haven't been able to find an instance where that happens - I've tried with really small and really large HTML files and wasn't able to get it to happen - I don't know if it's possible or not but I haven't seen it. I don't have a deep understanding of how the SAX pipeline breaks up its chunks

Reply:
have you tried logging "chunks" to see what they end up being?  I know it wouldnt be proof, but it might make us feel better if we see that it isnt just arbitrary length chunks that can end up including partial items.

Reply:
I've looked at chunks in the debugger and they really do appear to be pretty random. it sometimes seems to break on `\n`, but there are times when it will break apart a single HTML element in to multiple chunks. it seems to consistently put actual text between HTML tags into their own full chunks (even when it's a lot of text). I'm going to do some more testing around this to see if i can get any cases where either duplicate replacements happen or placeholders broken into multiple chunks and not correctly replaced

Reply:
ok, if the "body" of an HTML element is all one chunk (even if large) then I'm good with the original comment for this thread